p-ata/src/processor.rs

```diff
@@ -0,0 +1,595 @@
+/// is available in the instruction data.
+///
+/// An alternative was considered that used a loop with `derive_address` +
+/// `is_off_curve` instead of `find_program_address`, but though it saved ~30 CUs
+/// (1%) when bump happened to be `255`, it added more CUs on average.
+///
+/// ## Returns
+///
+/// `(address, bump)` - The canonical PDA address and its bump seed
+#[inline(always)]
+pub(crate) fn derive_canonical_ata_pda(
+    wallet: &Pubkey,
+    token_program: &Pubkey,
+    mint: &Pubkey,
+    program_id: &Pubkey,
+) -> (Pubkey, u8) {
+    find_program_address(
+        &[wallet.as_ref(), token_program.as_ref(), mint.as_ref()],
+        program_id,
+    )
+}
+
+/// Check if the given program ID is SPL Token (not Token-2022)
+#[inline(always)]
+pub(crate) fn is_spl_token_program(program_id: &Pubkey) -> bool {
+    const SPL_TOKEN_PROGRAM_ID: Pubkey =
+        pinocchio_pubkey::pubkey!("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
+    *program_id == SPL_TOKEN_PROGRAM_ID
+}
+
+/// Check if account data represents an initialized token account.
+/// Mimics p-token's is_initialized_account check.
+///
+/// Safety: caller must ensure account_data.len() >= 109.
+#[inline(always)]
+pub(crate) unsafe fn is_initialized_account(account_data: &[u8]) -> bool {
+    // Token account state is at offset 108 (after mint, owner, amount, delegate fields)
+    // State: 0 = Uninitialized, 1 = Initialized, 2 = Frozen
+    account_data[108] != 0
+}
+
+/// Validate that account data represents a valid token account.
+/// Replicates Token-2022's GenericTokenAccount::valid_account_data checks.
+#[inline(always)]
+pub(crate) fn valid_token_account_data(account_data: &[u8]) -> bool {
+    // Regular Token account: exact length match and initialized
+    if account_data.len() == TokenAccount::LEN {
+        // SAFETY: TokenAccount::LEN is compile-ensured to be == 165
+        return unsafe { is_initialized_account(account_data) };
+    }
+
+    // Token-2022's GenericTokenAccount::valid_account_data assumes Multisig
+    // if account_data length is Multisig::LEN. Collisions are prevented by
+    // adding a byte if a token account happens to have the same length as
+    // Multisig::LEN.
+    if account_data.len() > TokenAccount::LEN {
+        if account_data.len() == Multisig::LEN {
+            return false;
+        }
+        // SAFETY: TokenAccount::LEN is compile-ensured to be == 165, and in
+        // this branch account_data.len > TokenAccount::LEN
+        if unsafe { is_initialized_account(account_data) } {
+            return account_data[TokenAccount::LEN] == ACCOUNTTYPE_ACCOUNT;
+        }
+    }
+
+    false
+}
+
+/// Get zero-copy mint reference from account info
+#[inline(always)]
+pub(crate) unsafe fn get_mint_unchecked(account: &AccountInfo) -> &Mint {
+    let mint_data_slice = account.borrow_data_unchecked();
+    &*(mint_data_slice.as_ptr() as *const Mint)
+}
```

Comment:
Be careful, this is very unsafe! Can you put a safety warning saying to at least check that it's a mint or that it has 82 bytes?